backend/src/middleware/auth.test.ts

```diff
@@ -0,0 +1,179 @@
+import jwt from "jsonwebtoken";
+import { UnauthorizedError, ExpressError } from "../expressError";
+import {
+  authenticateJWT,
+  ensureLoggedIn,
+  ensureAdmin,
+  ensureCorrectUserOrAdmin,
+} from "./auth";
+import { Request, Response, NextFunction } from "express";
+
+import { SECRET_KEY } from "../config";
+
+const testJwt = jwt.sign({ username: "test", isAdmin: false }, SECRET_KEY);
+const badJwt = jwt.sign({ username: "test", isAdmin: false }, "wrong");
+
+describe("authenticateJWT", function () {
+  test("works: via header", function () {
+    expect.assertions(2);
+    const req = {
+      headers: { authorization: `Bearer ${testJwt}` },
+    };
+    const res = { locals: {} };
+    const next = function (err: ExpressError) {
+      expect(err).toBeFalsy();
+    };
+    authenticateJWT(req as Request, res as Response, next as NextFunction);
```

Comment:
Issue was that `req`, `res`, and `next` were not properly being treated as express `Request`, etc. The solution is to type cast them as such so Typescript knows that they should be treated as such. 

https://javascript.plainenglish.io/how-to-unit-test-express-middleware-typescript-jest-c6a7ad166e74